graviton/blackbox.py

```diff
@@ -174,406 +196,386 @@ def _partial_encode_assert(
         assert mode in ExecutionMode, f"unknown mode {mode} of type {type(mode)}"
         is_app = mode == ExecutionMode.Application
 
         abi_argument_types: Optional[List[EncodingType]] = None
-        abi_return_type: Optional[abi.ABIType] = None
+        abi_return_type: Optional[Union[abi.ABIType, str]] = None
+
+        method: Optional[abi.Method]
         if abi_method_signature:
-            """
-            Try to do the right thing.
-            When `omit_method_selector is False`:
-                * if provided with the same number of args as expected arg types
-                    --> prepend `None` to the types and `selector` to args
-                * if provided with |arg types| + 1 args
-                    --> assert that `args[0] == selector`
-                * otherwise
-                    --> there is a cardinality mismatch, so fail
-            When `omit_method_selector is True`:
-                * if provided with the same number of args as expected arg types
-                    --> good to go
-                * if provided with |arg types| + 1 args
-                    --> assert that `args[0] == selector` but DROP it from the args
-                * otherwise
-                    --> there is a cardinality mismatch, so fail
-            """
             method = abi.Method.from_signature(abi_method_signature)
             selector = method.get_selector()
             abi_argument_types = [a.type for a in method.args]
 
-            if validation:
-                args = list(args)
-                if len(args) == len(abi_argument_types):
-                    if not omit_method_selector:
-                        # the method selector is not abi-encoded, hence its abi-type is set to None
-                        abi_argument_types = [None] + abi_argument_types  # type: ignore
-                        args = [selector] + args
-
-                elif len(args) == len(abi_argument_types) + 1:
-                    assert (
-                        args[0] == selector
-                    ), f"{args[0]=} should have been the {selector=}"
-
-                    if omit_method_selector:
-                        args = args[1:]
-                    else:
-                        abi_argument_types = [None] + abi_argument_types  # type: ignore
-
-                else:
-                    raise AssertionError(
-                        f"{len(args)=} is incompatible with {len(method.args)=}: LEFT should be equal or exactly RIGHT + 1"
-                    )
-            elif not omit_method_selector:
-                abi_argument_types = [None] + abi_argument_types  # type: ignore
-
-            args = tuple(args)
-
             if method.returns.type != abi.Returns.VOID:
-                abi_return_type = cast(abi.ABIType, method.returns.type)
+                abi_return_type = method.returns.type
+        else:  # logic sigs always land here:
+            method = selector = None
```

Comment:
wut is this? `method = None` and `selector = None`?

Reply:
Perhaps this could be organized a bit better. But the `DryRunExecutor` handles all kinds of executions some with ABI types and some without. This comment assumes that logic sigs don't have any methods (which is basically true, but not really saying much). Look out for a small refactoring of this code coming soon.

Reply:
I believe this is an improvement: 17fd8acd87b04ee970c74b3728745e165b74ddf0

Reply:
lgtm